### PR TITLE
FOUNDATIONS: Add Group Post

### LINKED ISSUE #240: FOUNDATIONS: Add Group Post

### PR DESCRIPTION
**Pull request by ysnarafat**
closes #240

### CODE REVIEW

File: Taarafo.Core.Tests.Unit/Services/Foundations/GroupPosts/GroupPostServiceTests.Exception.Add.cs [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+using System.Threading.Tasks;
+using Microsoft.Data.SqlClient;
+using Moq;
+using Taarafo.Core.Models.GroupPosts.Exceptions;
+using Taarafo.Core.Models.GroupPosts;
+using Xunit;
+
+namespace Taarafo.Core.Tests.Unit.Services.Foundations.GroupPosts
+{
+    public partial class GroupPostServiceTests
```
Comment (ElbekDeveloper) [inline Taarafo.Core.Tests.Unit/Services/Foundations/GroupPosts/GroupPostServiceTests.Exception.Add.cs:19]: Name should be in plural:
Yours: `GroupPostServiceTests.Exception.Add.cs`
Desired: `GroupPostServiceTests.Exceptions.Add.cs`

File: Taarafo.Core.Tests.Unit/Services/Foundations/GroupPosts/GroupPostServiceTests.Exception.Add.cs [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+    {
+        [Fact]
+        public async Task ShouldThrowCriticalDependencyExceptionOnAddIfSqlErrorOccursAndLogItAsync()
```
Comment (ElbekDeveloper) [inline Taarafo.Core.Tests.Unit/Services/Foundations/GroupPosts/GroupPostServiceTests.Exception.Add.cs:22]: You need to handle a couple more exceptions that EF might throw

File: Taarafo.Core/Migrations/20220910032910_UpdateEventModel.cs [hunk 3]
```diff
@@ -1,4 +1,9 @@
-﻿using Microsoft.EntityFrameworkCore.Migrations;
+﻿// ---------------------------------------------------------------
```
Comment (ElbekDeveloper) [inline Taarafo.Core/Migrations/20220910032910_UpdateEventModel.cs:1]: Good practice would be not to touch any other file that is not in the scope of the issue.

File: Taarafo.Core/Services/Foundations/GroupPosts/GroupPostService.Validations.cs [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+{
+    public partial class GroupPostService
+    {
+        public static void ValidateGroupPostOnAdd(GroupPost groupPost)
+        {
+            ValidateGroupPostIsNotNull(groupPost);
+
+            Validate(
+                (Rule: IsInvalid(groupPost.GroupId), Parameter: nameof(GroupPost.GroupId)),
+                (Rule: IsInvalid(groupPost.Group), Parameter: nameof(GroupPost.Group)),
```
Comment (cjdutoit) [inline Taarafo.Core/Services/Foundations/GroupPosts/GroupPostService.Validations.cs:20]: Remove this line you only need to validate GroupId and PostId

File: Taarafo.Core/Services/Foundations/GroupPosts/GroupPostService.Validations.cs [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+                (Rule: IsInvalid(groupPost.PostId), Parameter: nameof(GroupPost.PostId)),
+                (Rule: IsInvalid(groupPost.Post), Parameter: nameof(GroupPost.Post)));
```
Comment (cjdutoit) [inline Taarafo.Core/Services/Foundations/GroupPosts/GroupPostService.Validations.cs:22]: Remove this line you only need to validate GroupId and PostId

File: Taarafo.Core.Tests.Unit/Services/Foundations/GroupPosts/GroupPostServiceTests.Validations.Add.cs [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+            };
+
+            var invalidGrouPostException = 
+                new InvalidGroupPostException();
+
+            invalidGrouPostException.AddData(
+                key: nameof(GroupPost.GroupId),
+                values: "Id is required");
+
+            invalidGrouPostException.AddData(
```
Comment (cjdutoit) [inline Taarafo.Core.Tests.Unit/Services/Foundations/GroupPosts/GroupPostServiceTests.Validations.Add.cs:72]: Remove this, you only need to validate GroupId and PostId

File: Taarafo.Core.Tests.Unit/Services/Foundations/GroupPosts/GroupPostServiceTests.Validations.Add.cs [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+                key: nameof(GroupPost.Group),
+                values: "Object is required");
+
+            invalidGrouPostException.AddData(
+                key: nameof(GroupPost.PostId),
+                values: "Id is required");
+
+            invalidGrouPostException.AddData(
```
Comment (cjdutoit) [inline Taarafo.Core.Tests.Unit/Services/Foundations/GroupPosts/GroupPostServiceTests.Validations.Add.cs:80]: Remove this, you only need to validate GroupId and PostId

File: Taarafo.Core.Tests.Unit/Services/Foundations/GroupPosts/GroupPostServiceTests.cs [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+
+        private static Expression<Func<Xeption, bool>> SameExceptionAs(Xeption expectedException) =>
+            actualException => actualException.SameExceptionAs(expectedException);
+
+        private static Filler<GroupPost> CreateGroupPostFiller()
+        {
+            var filler = new Filler<GroupPost>();
+
+            filler.Setup()
+                .OnType<DateTimeOffset>().IgnoreIt();
```
Comment (cjdutoit) [inline Taarafo.Core.Tests.Unit/Services/Foundations/GroupPosts/GroupPostServiceTests.cs:59]: Add an ignore rule for Group
Add an ignore rule for Post

**Review by cjdutoit — CHANGES_REQUESTED** [review]
Some small corrections needed.

**Comment by ElbekDeveloper** [comment]
This PR is closed due to inactivity

**Comment by ysnarafat** [comment]
> This PR is closed due to inactivity

Sorry for the late activity. I have addressed the comments and fixed accordingly. Could you please open the PR again and review the PR?

File: Taarafo.Core.Tests.Unit/Services/Foundations/GroupPosts/GroupPostServiceTests.Exceptions.Add.cs [hunk 9]
```diff
@@ -0,0 +1,236 @@
...
+            this.loggingBrokerMock.Verify(broker =>
+                broker.LogError(It.Is(SameExceptionAs(
+                    expectedGroupPostDependencyValidationException))),
+                        Times.Once);
+
+            this.storageBrokerMock.Verify(broker =>
+                broker.InsertGroupPostAsync(someGroupPost),
+                    Times.Once);
+
+            this.loggingBrokerMock.VerifyNoOtherCalls();
```
Comment (ElbekDeveloper) [inline Taarafo.Core.Tests.Unit/Services/Foundations/GroupPosts/GroupPostServiceTests.Exceptions.Add.cs:100]: There is one rule when verifying dependencies. You order these verification statements on the order by which they are called. 
In this case you first call `storageBroker` that throws exception and then `loggingBroker` logs the error. The order of these statements should be the same. This is the correct order: 

``` cs
this.storageBrokerMock.Verify(broker =>
    broker.InsertGroupPostAsync(someGroupPost),
        Times.Once);

this.loggingBrokerMock.Verify(broker =>
    broker.LogError(It.Is(SameExceptionAs(
            expectedGroupPostDependencyValidationException))),
                Times.Once);

this.storageBrokerMock.VerifyNoOtherCalls();
this.loggingBrokerMock.VerifyNoOtherCalls();

```

Comment (ysnarafat) [reply Taarafo.Core.Tests.Unit/Services/Foundations/GroupPosts/GroupPostServiceTests.Exceptions.Add.cs:100]: Okay, correcting according to this way.

**Review by ElbekDeveloper — APPROVED** [review]

**Review by cjdutoit — APPROVED** [review]

**Pull request merged**